x-pack/metricbeat/module/panw/vpn/gp_stats.go

```diff
@@ -0,0 +1,74 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package vpn
+
+import (
+	"encoding/xml"
+	"fmt"
+	"time"
+
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/x-pack/metricbeat/module/panw"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const gpStatsQuery = "<show><global-protect-gateway><statistics></statistics></global-protect-gateway></show>"
+
+func getGlobalProtectStatsEvents(m *MetricSet) ([]mb.Event, error) {
+
+	var response GPStatsResponse
+
+	output, err := m.client.Op(gpStatsQuery, panw.Vsys, nil, nil)
+	if err != nil {
+		m.logger.Error("Error: %s", err)
+		return nil, fmt.Errorf("error querying GlobalProtect statistics: %w", err)
+	}
+
+	err = xml.Unmarshal(output, &response)
+	if err != nil {
+		m.logger.Error("Error: %s", err)
+		return nil, fmt.Errorf("error unmarshaling GlobalProtect statistics response: %w", err)
+	}
+
+	events := formatGPStatsEvents(m, response)
+
+	return events, nil
+
+}
+
+func formatGPStatsEvents(m *MetricSet, response GPStatsResponse) []mb.Event {
+
+	if len(response.Result.Gateways) == 0 {
+		return nil
+	}
+
+	events := make([]mb.Event, 0, len(response.Result.Gateways))
+	timestamp := time.Now()
+
+	totalCurrent := response.Result.TotalCurrentUsers
+	totalPrevious := response.Result.TotalPreviousUsers
+
+	for _, gateway := range response.Result.Gateways {
+		event := mb.Event{
+			Timestamp: timestamp,
+			MetricSetFields: mapstr.M{
+				"globalprotect.gateway.name":           gateway.Name,
+				"globalprotect.gateway.current_users":  gateway.CurrentUsers,
+				"globalprotect.gateway.previous_users": gateway.PreviousUsers,
+				"globalprotect.total_current_users":    totalCurrent,
+				"globalprotect.total_previous_users":   totalPrevious,
+			},
+			RootFields: mapstr.M{
+				"observer.ip":     m.config.HostIp,
+				"host.ip":         m.config.HostIp,
+				"observer.vendor": "Palo Alto",
+				"observer.type":   "firewall",
+			}}
```

Comment:
I implemented @agithomas suggestion:
`RootFields: panw.MakeRootFields(m.config.HostIp)`

Reply:
Got it. That's fine for now. We can consider optimizations in the future, not at this moment. You can mark this comment as resolved.

However, using `panw.MakeRootFields(m.config.HostIp)` would indeed help in reducing code repetition. In each iteration, it makes a call to the function, which returns a new mapstr.M. While this approach does reduce code duplication, the repeated allocations are still there.

On the other hand, instantiating a single mapstr.M with the required values and reusing it throughout the process would be more efficient. This approach avoids creating a new mapstr.M object in each iteration, potentially reducing allocations, etc.

Reply:
I get your point about allocations. The observer vendor and type are obviously static, but the value of host.ip will change (though I only have one device to test with currently). Did you have an implementation in mind that takes that into account? Thanks.

Reply:
Yes, one is to not make a global variable but rather make a variable local to the function as I initially mentioned in my first comment because m.config.HostIp is never going to change in the local scope. So, the same mapstr.M across iterations.

Reply:
Ah, yes of course. I was kind of forgetting this was called in a loop :-) I will make that change.